firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
+                iter_stack.push(IterationNode::Visited {
+                    key: matched_key_nibbles.clone().into_boxed_slice(),
+                    children_iter: Box::new(
+                        get_children_iter(branch).filter(move |(_, pos)| *pos > nib),
+                    ),
+                });
+
+                // Figure out if the child at `nib` is a prefix of `key`.
+                // (i.e. if we should run this loop body again)
+                #[allow(clippy::indexing_slicing)]
+                let child_addr = match branch.children[nib as usize] {
+                    Some(c) => c,
+                    None => {
+                        // There is no child at `nib`.
+                        // We'll visit `node`'s first child at index > `nib`
+                        // first (if it exists).
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
                 };
 
-                self.poll_next(_cx)
-            }
+                matched_key_nibbles.push(nib);
+
+                let child = merkle
+                    .get_node(child_addr)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-            IteratorState::Iterating {
-                check_child_nibble,
-                visited_node_path,
-            } => {
-                let next = find_next_result(merkle, visited_node_path, check_child_nibble)
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))
-                    .transpose();
+                let partial_key = match child.inner() {
+                    NodeType::Branch(branch) => &branch.path,
+                    NodeType::Leaf(leaf) => &leaf.path,
+                    NodeType::Extension(_) => {
+                        panic!("extension nodes shouldn't exist")
+                    }
+                };
 
-                Poll::Ready(next)
+                let (comparison, new_unmatched_key_nibbles) =
+                    is_prefix(partial_key.iter(), unmatched_key_nibbles);
+                unmatched_key_nibbles = new_unmatched_key_nibbles;
+
+                match comparison {
+                    Ordering::Less => {
+                        // `child` is before `key`.
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
+                    Ordering::Equal => {
+                        // `child` is a prefix of `key`.
+                        matched_key_nibbles.extend(partial_key.iter().copied());
+                        node = child;
+                        node_addr = child_addr;
+                    }
+                    Ordering::Greater => {
+                        // `child` is after `key`.
+                        let child = merkle
+                            .get_node(child_addr)
+                            .map_err(|e| api::Error::InternalError(Box::new(e)))?;
```

Comment:
🤔 , did `child` or `child_addr` change here? Don't you already have the `child`?

Reply:
removed